gldemo/plugin.cpp

```diff
@@ -185,7 +184,7 @@ class gldemo : public threadloop {
 			glFlush();
 
 			// Publish our submitted frame handle to Switchboard!
-			auto frame = new rendered_frame;
+			rendered_frame* frame = new (_m_eyebuffer.allocate()) rendered_frame;
```

Comment:
This is nice and dandy but not everyone may be familiar with placement new. Plus, the example in the comments in `switchboard.hpp` doesn't do it this way -- it does allocate first, then populates, then puts. I think being consistent would make it easy for new users.
This is applicable to all the plugins (I can mark all the places if you wish).

Reply:
Placement new is not exactly equivalent to allocate-then-populate in the case where the struct has `const` fields which can't be set. I can use allocate-then-populate for all of the structs which don't have `const` fields, however.

Reply:
Switchboard needs to control the allocation of memory for messages for two reasons:
- We can use a "pool/slab" allocator (reusing the memory behind old events).
- We can use Switchboard between processes, if the memory behind the message comes from `mmap`ed shared memory.
Thus, I do not believe we can use the normal `new`.

Allocating and the mutating (as in the example) is not ideal because it won't work if the class has `const T` or `T&` members, it is possible to "forget" to set members, and all the messages require default-constructors.

There is one alternative to placement `new` however. We can have `writer.allocate(...)` forward its arguments to the constructor. This is how `std::make_shared<T>(...)` works. Unfortunately, it does not permit naming the parameters (aka designated initializers).

```suggestion
			switchboard::ptr<rendered_frame> frame = _m_eyebuffer.allocate(a, b, c, ..., z);
```

I think explicitly checking that all members have been passed is more important than that the programmer got the right position, because the type often implicitly checks the position.

Reply:
Bump

Reply:
I like this. It's neat. And if the programmer wanted to, they could construct an object on the stack using designated initializers and then call `allocate2` with that object, right?

Also, just to make sure, the plan is to still call it just [allocate](https://github.com/ILLIXR/ILLIXR/pull/32/commits/4a6307db2924e3fb25c34398d4a3d7495cfe9d43#diff-5fee54f9e607f7371d98fdca5009caf430563387cb34fde91855b985255a2cbcR462), right?

Reply:
Yes, just `allocate(...)`, replacing the old one.

I do not recommend manual allocation as in,

```C++
switchboard::ptr<rendered_frame> frame = _m_eyebuffer.allocate(switchboard::ptr<rendered_frame>{new rendered_frame {
    .a = a,
    .b = b,
}});
```

but it would be functionally correct when the developer does not need switchboard allocated memory (i.e. not IPC switchboard, not slab allocator).

Reply:
I was thinking of creating an object on the stack, not the heap, before calling allocate. So something like:
```
rendered_frame rf{.a = a, .b =b};
switchboard::ptr<rendered_frame> frame = _m_eyebuffer.allocate(rf);
```

Or is this not valid code?

Reply:
Oh, you're right. I've edited in place.

Reply:
Great. We're in agreement. Once this change is made, this PR is ready to go.